include/boost/geometry/algorithms/visit.hpp

```diff
@@ -0,0 +1,307 @@
+        std::enable_if_t<util::is_geometry_collection<Geom>::value, int> = 0
+    >
+    static bool visit_or_enqueue(F &, Geom &, std::deque<Iterator> & queue, Iterator iter)
+    {
+        queue.push_back(iter);
+        return true;
+    }
+    template
+    <
+        typename F, typename Geom, typename Iterator,
+        std::enable_if_t<! util::is_geometry_collection<Geom>::value, int> = 0
+    >
+    static bool visit_or_enqueue(F & f, Geom & g, std::deque<Iterator> & , Iterator)
+    {
+        return f(g);
+    }
+
+    template
+    <
+        typename Geom, typename Iterator,
+        std::enable_if_t<util::is_geometry_collection<Geom>::value, int> = 0
+    >
+    static void set_iterators(Geom & g, Iterator & first, Iterator & last)
+    {
+        first = boost::begin(g);
+        last = boost::end(g);
+    }
+    template
+    <
+        typename Geom, typename Iterator,
+        std::enable_if_t<!util::is_geometry_collection<Geom>::value, int> = 0
+    >
+    static void set_iterators(Geom &, Iterator &, Iterator &)
+    {}
+};
+
+
+} // namespace dispatch
+#endif // DOXYGEN_NO_DISPATCH
+
+
+// This function mimics std::visit
+// This algorithm treats GeometryCollection as StaticGeometry and calls function
+//   for it.
+// TODO: We could change the order of the arguments but I'm not sure if that's a good idea.
+template <typename UnaryFunction, typename Geometry>
+inline void visit(UnaryFunction && function, Geometry & geometry)
+{
+    dispatch::visit_one
+        <
+            Geometry
+        >::apply(std::forward<UnaryFunction>(function), geometry);
+}
+
+
+// This function mimics std::visit
+// This algorithm treats GeometryCollection as StaticGeometry and calls function
+//   for it.
+// TODO: We could change the order of the arguments but I'm not sure if that's a good idea.
+template <typename UnaryFunction, typename Geometry1, typename Geometry2>
+inline void visit(UnaryFunction && function, Geometry1 & geometry1, Geometry2 & geometry2)
+{
+    dispatch::visit_two
+        <
+            Geometry1, Geometry2
+        >::apply(std::forward<UnaryFunction>(function), geometry1, geometry2);
+}
+
+
+// This function works like combined std::visit and recursive std::for_each.
+// This algorithm works for all Geometries. It visits DynamicGeometries and calls function for
+//   StaticGeometries.
+// The number of elements visited may be different than the size of the top-most GeometryCollection
+//   if it's recursive.
+// TODO: This algorithm is similar to std::for_each so it could be called e.g. std::visit_each or
+//   std::for_each_visit but it doesn't work the same way, i.e. it traverses recursive
+//   GeometryCollections.
+// TODO: Should this algorithm work only for GeometryCollection?
+// TODO: We could change the order of the arguments but I'm not sure if that's a good idea.
+template <typename UnaryFunction, typename Geometry>
+inline void visit_breadth_first(UnaryFunction function, Geometry & geometry)
```

Comment:
Is this a function for the end user?
If so we should resolve all todo's and make a decision about the order.
Or should we move it to `namespace detail` first?

Reply:
Same for the `visit` functions above

Reply:
Yes, moving it to `detail` is a good idea regardless of the order of arguments.

Regarding the order, it depends if we think it's more like `std::visit` or `std::for_each`. I'm leaning more towards having the same interface for the whole family of `visit` functions also consistent with `std::visit`.

Reply:
I moved them to `detail`.